### DISCUSSION TITLE
Using timeline variables with the survey-likert plugin

### DISCUSSION THREAD

**Discussion by psychNerdJae** [jspsych/jsPsych#2619] [Q&A] [answered]
On each trial of my study, I'd like to present a likert prompt that changes according to the values inside a timeline variable. To provide a toy example: on trial one, I might ask how much participants like ice cream, and on trial two, how much they like veggies.

Naively, I'd expect the following code to accomplish this.

```javascript
var test = {
    timeline: [
        {
            type: jsPsychSurveyLikert,
            questions: [
                {
                    prompt: "How much do you like " + jsPsych.timelineVariable("test_prompt") + "?",
                    name: "like",
                    labels: ["Not at all", "Somewhat", "A great deal"]
                },
                {
                    prompt: "Is " + jsPsych.timelineVariable("test_prompt") + " your favorite food?",
                    name: "favorite",
                    labels: ["No", "Yes"]
                }
            ],
        }
    ],
    timeline_variables: [
        { test_prompt: "ice cream" },
        { test_prompt: "veggies" }
    ]
}
```

However, I am instead finding that the timeline variable seems to take on the value `[object Object]`, as can be seen in the screenshot below. It *does* loop over two trials, as expected, but displays the same prompt for both trials.

![jspsych_likert](https://user-images.githubusercontent.com/48289808/167337292-61af3b8e-9033-4536-b426-79de32cddfe6.png)

If anyone has pointers on how I might fix this, it'd be greatly appreciated!

**Comment by jodeleeuw** [comment, marked as answer]
Hi @psychNerdJae,

Because you are combining the timeline variable with other information you'll need to use a [dynamic parameter](https://www.jspsych.org/7.2/overview/dynamic-parameters/). 

```js
prompt: () => { return "How much do you like " + jsPsych.timelineVariable("test_prompt") + "?"; },
```

When you call `jsPsych.timelineVariable()` it actually returns a special kind of object, which jsPsych then treats like a dynamic parameter later on during the experiment run. So you are seeing `[object Object]` because the value of `jsPsych.timelineVariable()` is this object. When you wrap the whole `prompt` parameter in a function, then the jsPsych engine will wait to evaluate this timelineVariable until runtime and it will return the value of the timeline variable at that point.

Also, if you try this and it doesn't work, it might be because we haven't properly implemented support for nested parameters as timeline variables. I think we fixed this in a recent release, so I think it will work, but do follow up here if it doesn't before spending too much time trying to debug what might be our fault.

**Reply by psychNerdJae** [reply]
Hi @jodeleeuw, thanks for the quick response! Looks like the task is now working as expected. Somewhat embarrassingly, it looks like the page on dynamic parameters actually includes an example of using anonymous functions in the context of the `survey-x` family of plugins, in the subsection on nested parameters. I must have somehow missed that documentation.

Thanks also for your conceptual explanation for why this behavior emerges – it's very helpful to get an intuition for what I did wrong. It does appear that if I simply request `prompt: jsPsych.timelineVariable("test_prompt")` without adding any other text, jsPsych is more than happy to display the words "ice cream".
